### PR TITLE
5933 fixing loopy imports

### PR DESCRIPTION
**Pull request by harshi922**
[#5933](https://github.com/polyapi/poly-alpha/issues/5933)

Okay this is a big change but this makes it possible to not define types in CFX 3X. Tested against actual GHA CFX - all ok! Will work in server func logic similarly once this gets changes suggested and resolved!

### CODE REVIEW

File: polyapi/generate.py [hunk 1]
```diff
@@ -397,7 +403,16 @@ def clear() -> None:
     print("Cleared!")
 
 
-def render_spec(spec: SpecificationDto) -> Tuple[str, str]:
+def render_spec(spec: SpecificationDto) -> Tuple[str, str, str]:
```
Comment (eneumann) [inline polyapi/generate.py]: With this return-contract change, please make sure all callers can handle it (e.g. correct unpacking for all three values) This one in `rendered_spec.py` will break -> `func_str, type_defs = render_spec(spec)`

File: polyapi/client.py [hunk 2]
```diff
@@ -10,35 +13,247 @@
...
+            if isinstance(node, ast.TypeAlias) and isinstance(node.name, ast.Name):
+                class_names.add(node.name.id)
+
+    # Phase 3: Compute transitive dependencies of all classes
+    # Any module-level symbol that a class references (directly or transitively)
+    # must also be at module scope
+    module_scope_names: set[str] = set(class_names)
+
+    # Get all module-level assigned symbol names for reference
+    module_level_symbols: set[str] = {
```
Comment (eneumann) [inline polyapi/client.py]: I think there’s one runtime edge case here: class definitions can be moved to module scope before their imported dependencies are available. For example, with import pydantic + class A(pydantic.BaseModel), the class may be emitted outside the try, while import pydantic stays inside it. Can we treat imported symbols as module-scope dependencies too, so classes and their required imports stay aligned?

**Review by eneumann — COMMENTED** [review]
Nice work on this refactor @harshi922! 
A few of these edge cases look like the kind that static typing catches early (especially around optional TypedDict keys and changed return contracts). It may be worth adding a stronger type-checking extension to your vscode so you can spot these early on.
For example, my user settings (global settings) for python uses Pyright via Pylance in basic mode. Then most projects i work with also use mypy.

File: polyapi/generate.py (hunk 1 above)
Comment (harshi922) [reply polyapi/generate.py:406]: fixed

File: polyapi/client.py [hunk 3]
```diff
@@ -10,35 +13,247 @@
...
-    return "".join([prefix, code, "\n", suffix])
+    # Build try/except section for runtime code
+    prefix = f'logger = logging.getLogger("poly")\ntry:\n    '
+    suffix = (
+        f"\nexcept ImportError as e:\n"
+        f"    logger.warning(\"Failed to import client function "
+        f"'{function_name}', function unavailable: \" + str(e))"
+    )
+
+    indented = '\n    '.join(runtime_code.split('\n'))
```
[the remark this answers is not in the document]
Comment (harshi922) [reply polyapi/client.py]: pass fallback inplace

File: polyapi/client.py (hunk 2 above)
Comment (harshi922) [reply polyapi/client.py:405]: Makes sense! Got some test cases in too - pls let me know if it's missing anything happy to do another pass

File: polyapi/client.py [hunk 4]
```diff
@@ -10,35 +13,247 @@
...
+
+    This is the ONE narrow heuristic we still need because symtable
+    can't distinguish `X = Literal["a"]` (type alias) from `x = foo()` (runtime).
+
+    We check the VALUE, not the name — much more reliable than naming conventions.
     """
-    prefix = """logger = logging.getLogger("poly")
-try:
+    # X = Literal[...], X = Dict[str, Any], X = list[Foo], X = Union[...]
+    if isinstance(node, ast.Subscript):
```
[the remark this answers is not in the document]
Comment (harshi922) [reply polyapi/client.py:318]: done

File: polyapi/generate.py [hunk 5]
```diff
@@ -424,97 +439,129 @@ def _flatten_description(value: Any) -> List[str]:
...
-        func_str, func_type_defs = render_spec(spec)
+        module_scope_types, func_str, func_type_defs = render_spec(spec)
 
         if not func_str:
-            # If render_spec failed and returned empty string, don't create any files
             raise Exception("Function rendering failed - empty function string returned")
 
-        # Prepare all content first before writing any files
+        # Deduplicate types against previously emitted types for this context
+        unique_types = _deduplicate_types(full_path, module_scope_types)
```
Comment (eneumann) [inline polyapi/generate.py:561]: `unique_types` is never used, which means the extracted client imports/types are still dropped from the generated module and custom functions can fail at import time.

File: polyapi/client.py [hunk 6]
```diff
@@ -1,44 +1,302 @@
...
+    This is the ONE narrow heuristic we still need because symtable
+    can't distinguish `X = Literal["a"]` (type alias) from `x = foo()` (runtime).
+
+    We check the VALUE, not the name — much more reliable than naming conventions.
+    """
+    # X = Literal[...], X = Dict[str, Any], X = list[Foo], X = Union[...]
+    if isinstance(node, ast.Subscript) and isinstance(node.value, ast.Name):
+        return node.value.id in _TYPING_SUBSCRIPT_NAMES
+    # X = str | int | float new Union
+    if isinstance(node, ast.BinOp) and isinstance(node.op, ast.BitOr):
```
Comment (eneumann) [inline polyapi/client.py:340]: This still looks too broad: regular runtime expressions like `FLAGS = 1 | 2` would also be classified as type aliases here.

**Review by eneumann — COMMENTED** [review]
Nice progress here — most of my earlier comments look addressed. I still see two remaining issues to tighten up and then this should be g2g!

File: polyapi/client.py (hunk 6 above)
Comment (harshi922) [reply polyapi/client.py:340]: I think I got this one but I cannot figure out how to get MyUnion = Foo | Bar T.T since neither will be in _BUILTIN_TYPE_NAMES and so gets missed. Any suggestions ?

**Comment by harshi922** [comment]
@eneumann 

Current state: 
 
## missed hoist cases are:

(type alias stays inside try/except)

#### Aliased typing_extensions ( Guess we can throw warning
import typing_extensions as t
X = t.Optional[int]      # 't' not in _TYPING_MODULES - MISSED
T = t.TypeVar("T")       # same

-----------------------------------------------------------------------

#### Two custom class union without None
MyUnion = Foo | Bar      # neither in _BUILTIN_TYPE_NAMES - MISSED

## Wrong hoist cases are: 

#### Shadowed typing name as subscript
List = get_runtime_list()
X = List[0]   # 'List' is in _TYPING_SUBSCRIPT_NAMES - hoisted wrongly

-----------------------------------------------------------------------

#### User function that shadows a typing functional name
def TypeVar(name): return name
T = TypeVar("T")   # func.id in _TYPING_FUNCTIONAL_NAMES - wrongly hoisted

So, do you think it makes sense if I grab https://github.com/polyapi/poly-alpha/issues/4073 and throw warns on above missed cases and we refuse to deploy cfxs that have the og 35 keywords + wtv else lib imports either as part of this or a new ticket ?

File: polyapi/client.py [hunk 7]
```diff
@@ -1,44 +1,431 @@
...
+
+        is_type_import = False
+        is_type_def = False
+
+        # Imports: safe typing/stdlib imports go to module scope;
+        # also promote any import that binds a name required by a module-scope class.
+        if isinstance(node, (ast.Import, ast.ImportFrom)):
+            is_type_import = (
+                _is_safe_import(node)
+                or bool(_import_bound_names(node) & module_scope_names)
```
Comment (eneumann) [inline polyapi/client.py]: This moves third-party imports required by hoisted type definitions outside the generated client import guard. If that dependency is missing, the whole generated context can fail to import before this function can be marked unavailable.

File: polyapi/client.py [hunk 8]
```diff
@@ -1,44 +1,431 @@
...
+    # so this is the ONE remaining heuristic — but scoped narrowly to
+    # assignments whose RHS is a typing construct (Subscript/BinOp with |)
+    for node in ast.iter_child_nodes(tree):
+        if not isinstance(node, ast.Assign):
+            continue
+        new_names = [
+            t.id for t in node.targets
+            if isinstance(t, ast.Name) and t.id not in module_scope_names
+        ]
+        if new_names and _rhs_is_type_construct(node.value):
```
Comment (eneumann) [inline polyapi/client.py]: This hoists assignments based on typing-looking names without checking whether those names were shadowed locally. That can move runtime expressions before their bindings and make valid client code fail during generated module import.

**Review by eneumann — COMMENTED** [review]
Thanks @harshi922, just a few generated-code import-order issues to address.

**Comment by eneumann** [comment]
> @eneumann
> 
> Current state:
> 
> ## missed hoist cases are:
> (type alias stays inside try/except)
> 
> #### Aliased typing_extensions ( Guess we can throw warning
> import typing_extensions as t X = t.Optional[int] # 't' not in _TYPING_MODULES - MISSED T = t.TypeVar("T") # same
> 
> #### Two custom class union without None
> MyUnion = Foo | Bar # neither in _BUILTIN_TYPE_NAMES - MISSED
> 
> ## Wrong hoist cases are:
> #### Shadowed typing name as subscript
> List = get_runtime_list() X = List[0] # 'List' is in _TYPING_SUBSCRIPT_NAMES - hoisted wrongly
> 
> #### User function that shadows a typing functional name
> def TypeVar(name): return name T = TypeVar("T") # func.id in _TYPING_FUNCTIONAL_NAMES - wrongly hoisted
> 
> So, do you think it makes sense if I grab [polyapi/poly-alpha#4073](https://github.com/polyapi/poly-alpha/issues/4073) and throw warns on above missed cases and we refuse to deploy cfxs that have the og 35 keywords + wtv else lib imports either as part of this or a new ticket ?

@harshi922 I think #4073 is a good fit for blocking reserved words in Poly item names and context segments. I’d keep that separate from these code-hoisting cases though. Missed hoists seem fine as warnings/follow-up work, but wrong hoists should be fixed in this PR since they can generate broken Python. Blocking deploys based on local aliases or typing-looking symbols feels a bit too broad until the generator can track where those symbols actually came from.

File: polyapi/client.py [hunk 9]
```diff
@@ -1,44 +1,596 @@
...
+
+        # Imports: only safe (stdlib/typing) imports go to module scope.
+        # Non-safe imports stay in try/except — hoisting them risks crashing the whole
+        # generated module if the dependency is missing (before the guard can fire).
+        if isinstance(node, (ast.Import, ast.ImportFrom)):
+            is_type_import = _is_safe_import(node)
+
+        # Class definitions: symtable confirmed these are classes
+        elif isinstance(node, ast.ClassDef):
+            is_type_def = node.name in class_names  # always True, but explicit
```
Comment (eneumann) [inline polyapi/client.py]: This still leaves the class outside the guard but keeps its required import inside the guard. If that import is missing, the generated module can fail before the `ImportError` handling runs.

File: polyapi/client.py [hunk 10]
```diff
@@ -1,44 +1,596 @@
...
+        ]
+        if not new_names or not _rhs_is_type_construct(node.value, typing_bound, local_shadows):
+            continue
+
+        start = node.lineno - 1
+        end = getattr(node, 'end_lineno', None) or start + 1
+
+        unresolved = {
+            n.id for n in ast.walk(node.value)
+            if isinstance(n, ast.Name) and n.id not in all_known_names
```
Comment (eneumann) [inline polyapi/client.py]: This rewrites `typing` / `typing_extensions` itself into a string, so aliases like `typing.Optional[int]` can fail when the generated module imports.

File: polyapi/client.py [hunk 11]
```diff
@@ -1,44 +1,596 @@
...
+    This is the ONE narrow heuristic we still need because symtable
+    can't distinguish `X = Literal["a"]` (type alias) from `x = foo()` (runtime).
+
+    We check the VALUE, not the name. `typing_bound` extends the static sets with
+    names actually imported from typing/typing_extensions in the current file.
+    `local_shadows` removes names from the static sets that were redefined at module
+    scope (e.g. `List = []` or `def TypeVar(...)`), preventing wrong hoists.
+    """
+    all_subscript = (_TYPING_SUBSCRIPT_NAMES | typing_bound) - local_shadows
+    all_functional = (_TYPING_FUNCTIONAL_NAMES | typing_bound) - local_shadows
```
Comment (eneumann) [inline polyapi/client.py]: This can move real runtime code into the type-only section. For example, `typing.cast(...)` runs at runtime, but this logic may treat it like a type definition and move it before the values it needs exist.

**Review by eneumann — COMMENTED** [review]
Thanks for all the iteration on this @harshi922!
The shadowing cases look much better now, and this is getting a lot closer. I left a few remaining edge cases around import-time behavior, but the direction looks good and the recent changes addressed the big false-positive case.

File: polyapi/client.py (hunk 9 above)
Comment (harshi922) [reply polyapi/client.py]: class Foo(pydantic.BaseModel) stays inside try block with from pydantic import BaseModel

File: polyapi/client.py (hunk 10 above)
Comment (harshi922) [reply polyapi/client.py]: typing.Optional[int] not rewritten

File: polyapi/client.py (hunk 11 above)
Comment (harshi922) [reply polyapi/client.py]: only the explicit static list (TypeVar, TypedDict, NewType, ParamSpec, TypeVarTuple, TypeAliasType, ForwardRef, NamedTuple) is trusted for functional type forms

File: polyapi/client.py (hunk 6 above)
Comment (eneumann) [reply polyapi/client.py:340]: Yeah, don’t try to hoist `Foo | Bar` unless we already know `Foo` and `Bar` are module-scope type/class names. Missed hoists are safer than wrong hoists.

For this case, the main fix seems to be making the `|` logic respect the same shadow checks as the other type-alias paths, so a local `Optional` / `List` / `Literal` binding is not treated as a typing type.

**Review by eneumann — COMMENTED** [review]
Awesome! I think all that is left now is to fix the failing tests and tighten the remaining hoist edge case.

File: polyapi/client.py [hunk 12]
```diff
@@ -1,44 +1,660 @@
...
+
+    # Phase 4: Classify each AST node using the symtable results
+    type_import_lines: set[int] = set()
+    type_def_lines: set[int] = set()
+    modified_type_lines: dict[int, str] = {}  # line index -> replacement source
+
+    prev_was_type_def = False
+
+    for node in ast.iter_child_nodes(tree):
+        start = getattr(node, 'lineno', 1) - 1
```
Comment (eneumann) [inline polyapi/client.py:543]: We should avoid splitting decorated classes here. Right now the decorator line stays in the guarded block, which can make it decorate the next generated function or crash before the per-function import guard runs.

File: polyapi/client.py [hunk 13]
```diff
@@ -1,44 +1,660 @@
...
+            if isinstance(node.target, ast.Name):
+                rhs_names_available = all(
+                    n.id in module_scope_available
+                    for n in ast.walk(node.value)
+                    if isinstance(n, ast.Name)
+                )
+                is_type_def = node.target.id in module_scope_names and rhs_names_available
+
+        # Function definitions: NEVER module scope (these are runtime)
+        elif isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
```
Comment (eneumann) [inline polyapi/client.py:603]: Might need to re-check class dependencies that are local functions before hoisting the class? The generated module can evaluate the class before those names exist.

File: polyapi/client.py [hunk 14]
```diff
@@ -1,44 +1,660 @@
...
+    """Check if an import statement is safe to place at module scope.
+
+    Safe imports are stdlib and typing modules that will never raise ImportError.
+    """
+    if isinstance(node, ast.Import):
+        return all(
+            alias.name.split('.')[0] in SAFE_IMPORT_MODULES
+            for alias in node.names
+        )
+    if isinstance(node, ast.ImportFrom):
```
Comment (eneumann) [inline polyapi/client.py:110]: This could use one more guard for `from ... import ...` cases. The root module might be safe, but that specific imported name can be missing or unavailable in the generated runtime, which can cause import errors.

File: polyapi/client.py [hunk 15]
```diff
@@ -1,44 +1,660 @@
...
+    """Return a deep copy of node with unresolved Name refs replaced by string literals.
+
+    Produces forward references like Union["UnknownType", KnownType] so the
+    hoisted assignment is valid Python even when the name is defined later
+    (same-file) or in another module (cross-context).
+    """
+    node = copy.deepcopy(node)
+    for parent in ast.walk(node):
+        for field, val in ast.iter_fields(parent):
+            if isinstance(val, ast.Name) and val.id not in all_known:
```
Comment (eneumann) [inline polyapi/client.py]: Small edge case here -> when an unresolved name is the base of a dotted type ref, quoting it can turn the generated module-scope code into a string attribute lookup. That can raise before the import guard gets a chance to isolate just that function.

File: polyapi/client.py [hunk 16]
```diff
@@ -1,44 +1,660 @@
...
+        # AND that all RHS names are available at module scope. The Phase 3b
+        # cascade only follows class symtable deps — a non-class alias like
+        # STATUS = SomeType can land in module_scope_names via BFS without
+        # SomeType being checked, causing a NameError when hoisted.
+        elif isinstance(node, ast.Assign):
+            target_in_scope = any(
+                isinstance(t, ast.Name) and t.id in module_scope_names
+                for t in node.targets
+            )
+            rhs_names_available = all(
```
Comment (eneumann) [inline polyapi/client.py:585]: One more shadowing edge case -> the availability check can trust names that look like typing helpers even if this function rebinds them or imports them inside the guarded block. If a hoisted class depends on that alias, the generated module can evaluate it before the real binding exists.

File: polyapi/client.py [hunk 17]
```diff
@@ -1,44 +1,660 @@
...
+    If `List = get_runtime_list()` or `def TypeVar(...):` appears in the file,
+    those names must not be trusted as typing constructs in Phase 5 even though
+    they appear in the static whitelist.
     """
-    prefix = """logger = logging.getLogger("poly")
-try:
+    all_typing = _TYPING_SUBSCRIPT_NAMES | _TYPING_FUNCTIONAL_NAMES
+    shadowed: set[str] = set()
+    for node in ast.iter_child_nodes(tree):
+        if isinstance(node, ast.Assign):
```
Comment (eneumann) [inline polyapi/client.py:195]: Destructuring case to guard -> `Optional, other = ...` creates a local `Optional` value, but the alias hoist can treat `Optional` like the normal typing helper. That means `Alias = Optional[str]` can run before this function’s local `Optional` value exists.

File: polyapi/client.py [hunk 18]
```diff
@@ -1,44 +1,660 @@
...
+    # remaining heuristic — scoped narrowly to typing-construct RHS shapes.
+    #
+    # Unresolved names (defined later in the same file or in another module) are
+    # replaced with string forward references via _quote_unresolved_names so the
+    # hoisted line is always valid Python and mypy can resolve same-file refs.
+    all_known_names = _BUILTIN_TYPE_NAMES | typing_bound | module_scope_names
+    for node in ast.iter_child_nodes(tree):
+        if not isinstance(node, ast.Assign):
+            continue
+        new_names = [
```
Comment (eneumann) [inline polyapi/client.py:641]: If a whole chained assignment gets hoisted because one target is a type alias, the other targets on that line move too. If one of those needs runtime state, it can fail before the import guard runs.

File: polyapi/client.py [hunk 19]
```diff
@@ -1,44 +1,660 @@
...
+
+    # Build try/except section for runtime code
+    prefix = f'logger = logging.getLogger("poly")\ntry:\n    '
+    suffix = (
+        f"\nexcept ImportError as e:\n"
+        f"    logger.warning(\"Failed to import client function "
+        f"'{function_name}', function unavailable: \" + str(e))"
+    )
+
+    body = runtime_code if runtime_code.strip() else 'pass'
```
Comment (eneumann) [inline polyapi/client.py]: Type-only client code can leave just comments in the runtime block, resulting in an indentation error from an empty `try` body.

**Review by eneumann — COMMENTED** [review]
Thanks for all the hard work on this @harshi922! This is a big and tricky change, and I know there are a lot of weird Python import/type-hoisting cases to think through 🫠. The PR is definitely closer and I thought we were there last go around, but after another closer look, there are still more edge cases that can break the generated context before the per-function guard can help.

File: polyapi/client.py (hunk 19 above)
Comment (harshi922) [reply polyapi/client.py]: scan each line and require at least one that isn't blank and doesn't start with # before using runtime_code as the body; otherwise substitute 'pass'

File: polyapi/client.py (hunk 18 above)
Comment (harshi922) [reply polyapi/client.py:641]: changed any to all so the assignment is only hoisted if every single target is already confirmed to belong at module scope

File: polyapi/client.py (hunk 17 above)
Comment (harshi922) [reply polyapi/client.py:195]: subtract local_shadows from module_scope_available so rebound names don't pass the check

File: polyapi/client.py (hunk 16 above)
Comment (harshi922) [reply polyapi/client.py:585]: use ast.walk(t) instead, which traverses the whole target structure and finds Name nodes nested inside tuples/lists

File: polyapi/client.py (hunk 15 above)
Comment (harshi922) [reply polyapi/client.py]: _quote_expr to "some_module.SomeType" instead of just quoting the base

File: polyapi/client.py (hunk 14 above)
Comment (harshi922) [reply polyapi/client.py:110]: after confirming the module root is safe, actually call importlib.import_module(module) and check hasattr(mod, alias.name) for each imported name. If the name doesn't exist in the module - not safe to hoist

File: polyapi/client.py (hunk 13 above)
Comment (harshi922) [reply polyapi/client.py:603]: Method bodies are skipped because they only execute when called - not when the class is defined. Nested class bodies are still traversed because they execute at class-definition time

File: polyapi/client.py (hunk 12 above)
Comment (harshi922) [reply polyapi/client.py:543]: when hoisting a class that has decorators, walk back start to the first decorator's line number so all decorator lines travel with the class

File: polyapi/generate.py [hunk 20]
```diff
@@ -424,97 +439,129 @@ def _flatten_description(value: Any) -> List[str]:
...
+            target = node.targets[0]
+            if isinstance(target, ast.Name):
+                name = target.id
+        elif isinstance(node, ast.AnnAssign) and isinstance(node.target, ast.Name):
+            name = node.target.id
+        
+        if not name:
+            continue
+        
+        start = getattr(node, 'lineno', 1) - 1
```
Comment (eneumann) [inline polyapi/generate.py:524]: Thanks for fixing the decorated-class extraction path! I think the same decorator handling needs to happen in dedupe too. If the same decorated class is emitted again, _deduplicate_types() starts at the class line, so it can remove the duplicate class body but leave the decorator behind in __init__.py.

File: polyapi/client.py (hunk 16 above)
Comment (eneumann) [reply polyapi/client.py:585]: Ah got it, that fixes the destructuring shadow case! However, I think there’s still a related dependency split here. If a hoisted class depends on an assignment or decorator that stays inside the guarded block, the generated module can still evaluate the class before that binding exists.

Maybe simplest is to track the assignment names where `assignment_dependencies_available` is false, then run the same dependency cascade as the unsafe-import pruning so any class depending on those names stays guarded too.

File: polyapi/client.py [hunk 21]
```diff
@@ -1,44 +1,769 @@
...
+    # Also handles typing.Optional[int] where node.value is ast.Attribute
+    if isinstance(node, ast.Subscript):
+        val = node.value
+        head_ok = False
+        if isinstance(val, ast.Name):
+            head_ok = val.id in all_subscript
+        elif isinstance(val, ast.Attribute):
+            head_ok = (
+                isinstance(val.value, ast.Name)
+                and val.value.id in _TYPING_MODULES
```
Comment (eneumann) [inline polyapi/client.py:326]: One more tiny shadowing case, if client code rebinds `typing`, we can still hoist `typing.Optional[...]` before `typing` exists.

**Review by eneumann — COMMENTED** [review]
Thanks @harshi922! We are soooo close, just two import-time edge cases around hoisted code depending on bindings.

**Comment by eupharis** [comment]
@harshi922 no activity since Feb looks like! could we close this one for now until you have time to pick it up later?

...

oh wait, looks like review on May 6 nevermind! carry on!

**Comment by harshi922** [comment]
@eupharis yepp will resume work on this sisyphus ticket soon!
